discovery_client.go

```diff
@@ -0,0 +1,148 @@
+package rendezvous
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type rendezvousDiscoveryClient struct {
+	rp        RendezvousPoint
+	peerCache sync.Map //is a map[string]discoveredPeerCache
+	rng       *rand.Rand
+	rngMux    sync.Mutex
+}
+
+type discoveredPeerCache struct {
+	cachedRegs map[peer.ID]*Registration
+	cookie     []byte
+	mux        sync.Mutex
+}
+
+func NewRendezvousDiscoveryClient(host host.Host, rendezvousPeer peer.ID) discovery.Discovery {
+	rp := NewRendezvousPoint(host, rendezvousPeer)
+	return &rendezvousDiscoveryClient{rp, sync.Map{}, rand.New(rand.NewSource(rand.Int63())), sync.Mutex{}}
+}
+
+func (c *rendezvousDiscoveryClient) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return 0, err
+	}
+
+	ttl := options.Ttl
+	var ttlSeconds int
+
+	if ttl == 0 {
+		ttlSeconds = 7200
+	} else {
+		ttlSeconds = int(math.Round(ttl.Seconds()))
+	}
+
+	if returnedTTL, err := c.rp.Register(ctx, ns, ttlSeconds); err != nil {
+		return 0, err
+	} else {
+		return returnedTTL, nil
+	}
+}
+
+func (c *rendezvousDiscoveryClient) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	const maxLimit = 1000
+	limit := options.Limit
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
+	}
+
+	// Get cached peers
+	var cache *discoveredPeerCache
+
+	genericCache, _ := c.peerCache.LoadOrStore(ns, &discoveredPeerCache{})
+	cache = genericCache.(*discoveredPeerCache)
+
+	cache.mux.Lock()
+
+	// Remove all expired entries from cache
+	currentTime := int(time.Now().Unix())
+	newCacheSize := len(cache.cachedRegs)
+
+	for p := range cache.cachedRegs {
+		reg := cache.cachedRegs[p]
+		if reg.Ttl < currentTime {
+			newCacheSize--
+			delete(cache.cachedRegs, p)
+		}
+	}
+
+	cookie := cache.cookie
+	cache.mux.Unlock()
+
+	// Discover new records if we don't have enough
+	var discoveryErr error
```

Comment:
just `err`.

Reply:
If I name this `err` I need to rename the internal variable currently named `err` or we run into scoping problems. If you'd like to name both that's fine with me.

Reply:
why is this a different error?

Reply:
it's not a different error it's just an interaction between scoping and the way `=` and  `:=` work with multiple return values. I can just invert this and name the inner one `discoveryErr`, `terr`, or whatever and then replace `discoveryErr=err` with `err=discoveryErr` no var declaration required. This sufficient?

Reply:
you can declare a variable first before assigning, this will solve your scoping problem.

Reply:
Ok, I'll need two preassignment declarations (one for `records` and one for `newCookie`) but that's fine with me.

Reply:
struct copy is cheap, don't worry about it.